File: blockchain-info-aggregation-stream-parent/blockchain-info-aggregation-stream-infrastructure/src/test/java/info/blockchain/aggregation/infrastructure/stream/ReplaySubjectMessageStreamTest.java
package info.blockchain.aggregation.infrastructure.stream;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import reactor.core.publisher.Flux;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Stream;

import static org.awaitility.Awaitility.await;

public class ReplaySubjectMessageStreamTest {

    private ReplaySubjectMessageStream<String> stream;

    @Before
    public void setup() {
        this.stream = new ReplaySubjectMessageStream<>();
    }

    @Test
    public void should_replay_messages_to_observers() {

        final AtomicInteger signals = new AtomicInteger();
        Flux.from(stream.getSubscriber().publisher())
                .doOnNext(s -> signals.incrementAndGet())
                .subscribe();

        MessagePublisher<String> messagePublisher = stream.getPublisher();

        Stream.of("element1", "element2").forEach(messagePublisher::publish);

        await().timeout(500, TimeUnit.MILLISECONDS).until(() -> signals.get() == 2);

    }
}

File: blockchain-info-aggregation-stream-parent/blockchain-info-aggregation-stream-infrastructure/src/test/java/info/blockchain/aggregation/infrastructure/stream/ReplaySubjectMessageStreamTest.kt
package info.blockchain.aggregation.infrastructure.stream

import org.awaitility.Awaitility.await
import org.junit.Before
import org.junit.Test
import reactor.core.publisher.Flux
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import java.util.stream.Stream

class ReplaySubjectMessageStreamTest {

    private var stream: ReplaySubjectMessageStream<String>? = null

    @Before
    fun setup() {
        this.stream = ReplaySubjectMessageStream()
    }

    @Test
    fun should_replay_messages_to_observers() {

        val signals = AtomicInteger()
        Flux.from(stream!!.subscriber.publisher())
                .doOnNext { signals.incrementAndGet() }
                .subscribe()

        val messagePublisher = stream!!.publisher

        Stream.of("element1", "element2").forEach { messagePublisher.publish(it) }

        await().timeout(500, TimeUnit.MILLISECONDS).until { signals.get() == 2 }

    }
}
